modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionDemander.java

```diff
@@ -796,131 +797,148 @@ private void mvccPreloadEntries(
 
-                        int cacheId = entryHist.get(0).cacheId();
+                        if (flushHistory) {
+                            assert !entryHist.isEmpty();
 
-                        if (grp.sharedGroup() && (cctx == null || cacheId != cctx.cacheId())) {
-                            assert cacheId != CU.UNDEFINED_CACHE_ID;
+                            int cacheId = entryHist.get(0).cacheId();
 
-                            cctx = grp.shared().cacheContext(cacheId);
-                        }
+                            if (grp.sharedGroup() && (cctx == null || cacheId != cctx.cacheId())) {
+                                assert cacheId != CU.UNDEFINED_CACHE_ID;
+
+                                cctx = grp.shared().cacheContext(cacheId);
+                            }
+
+                            if (cctx != null) {
+                                if (!mvccPreloadEntry(cctx, node, entryHist, topVer, p)) {
+                                    if (log.isTraceEnabled()) {
+                                        log.trace("Got entries for invalid partition during " +
+                                            "preloading (will skip) [p=" + p +
+                                            ", entry=" + entryHist.get(entryHist.size() - 1) + ']');
+                                    }
+
+                                    return; // Skip current partition.
+                                }
 
-                        if (cctx != null) {
-                            if (!mvccPreloadEntry(cctx, node, entryHist, topVer, p)) {
-                                if (log.isTraceEnabled())
-                                    log.trace("Got entries for invalid partition during " +
-                                        "preloading (will skip) [p=" + p +
-                                        ", entry=" + entryHist.get(entryHist.size() - 1) + ']');
+                                rebalanceFut.onReceivedKeys(p, 1, node);
 
-                                return; // Skip current partition.
+                                receivedKeys.computeIfAbsent(cacheId, cid -> new AtomicLong()).incrementAndGet();
                             }
 
-                            rebalanceFut.onReceivedKeys(p, 1, node);
+                            if (!hasMore)
+                                return;
 
-                            updateGroupMetrics();
+                            entryHist.clear();
                         }
 
-                        if (!hasMore)
-                            return;
-
-                        entryHist.clear();
+                        entryHist.add(entry);
                     }
-
-                    entryHist.add(entry);
                 }
-            }
-            finally {
-                ctx.database().checkpointReadUnlock();
+                finally {
+                    ctx.database().checkpointReadUnlock();
+                }
             }
         }
+        finally {
```

Comment:
Do not use try-finaly block here.
If there is any exception has happened just miss an update in the metric.